Zend/zend_closures.c

```diff
@@ -39,6 +39,7 @@ typedef struct _zend_closure {
 	zval              this_ptr;
 	zend_class_entry *called_scope;
 	zif_handler       orig_internal_handler;
+	HashTable        *static_variables;
```

Comment:
I didn't completely understand, why do we need this new field. Just to reserve a new slot for map_ptr? And why  we can't still use ``closure->func.op_array.static_variables``?

Reply:
We need to avoid overwriting the original `static_variables` here, so they can still be freed (if the closure has the last reference to the function).

Reply:
I still didn't get this. We copy the whole op_array into zend_closure, so why can't we reuse ``closure->func.op_array.static_variables``?
I replaced usage of ``closure->static_variables`` by ``closure->func.op_array.static_variables`` in your patch and don't see any tests failures (in Zend tests ext/opcache). Please demonstrate the failure.

Reply:
> I still didn't get this. We copy the whole op_array into zend_closure, so why can't we reuse `closure->func.op_array.static_variables`?

The issue is that while we copy the op_array itself, it points to shared resources (including static_variables). These are dropped when the op_array refcount reaches zero. If we reuse static_variables for a different purpose and refcount drop to zero in the closure dtor, then we will end up leaking the original static_variables.

> I replaced usage of `closure->static_variables` by `closure->func.op_array.static_variables` in your patch and don't see any tests failures (in Zend tests ext/opcache). Please demonstrate the failure.

If I undo these changes using the following diff
```
diff --git a/Zend/zend_closures.c b/Zend/zend_closures.c
index b3ce89a30d..5e05da46c7 100644
--- a/Zend/zend_closures.c
+++ b/Zend/zend_closures.c
@@ -39,7 +39,6 @@ typedef struct _zend_closure {
        zval              this_ptr;
        zend_class_entry *called_scope;
        zif_handler       orig_internal_handler;
-       HashTable        *static_variables;
 } zend_closure;
 
 /* non-static since it needs to be referenced */
@@ -689,13 +688,11 @@ static void zend_create_closure_ex(zval *res, zend_function *func, zend_class_en
                /* For fake closures, we want to reuse the static variables of the original function. */
                if (!is_fake) {
                        if (closure->func.op_array.static_variables) {
-                               HashTable *static_variables =
-                                       ZEND_MAP_PTR_GET(closure->func.op_array.static_variables_ptr);
-                               closure->static_variables = zend_array_dup(
-                                       static_variables ? static_variables : closure->func.op_array.static_variables);
+                               closure->func.op_array.static_variables =
+                                       zend_array_dup(closure->func.op_array.static_variables);
                        }
                        ZEND_MAP_PTR_INIT(closure->func.op_array.static_variables_ptr,
-                               &closure->static_variables);
+                               &closure->func.op_array.static_variables);
                }
 
                /* Runtime cache is scope-dependent, so we cannot reuse it if the scope changed */
```
then I get 30 failing tests on Zend/ due to memory leaks (debug, no opcache).

Reply:
ah yes, I tested with opcache.

Reply:
All the failures are memory leaks, because of missed deallocation of op_array.static_variables of dynamic functions, that is guarded by op_array->refcount.

I understood, how the patch works, but this looks a bit weird. We allocate new pointer to keep old/unused value of op_array.static_variables just to destroy it.

Reply:
I think, it's more clear to explicitly deallocate static_variables of dynamic functions.

```diff
diff --git a/Zend/zend_opcode.c b/Zend/zend_opcode.c
index ac9cc5f704..54edd806f2 100644
--- a/Zend/zend_opcode.c
+++ b/Zend/zend_opcode.c
@@ -600,6 +607,16 @@ ZEND_API void destroy_op_array(zend_op_array *op_array)
        if (op_array->static_variables) {
                zend_array_destroy(op_array->static_variables);
        }
        if (op_array->num_dynamic_func_defs) {
                for (i = 0; i < op_array->num_dynamic_func_defs; i++) {
+                       if (op_array->dynamic_func_defs[i]->static_variables) {
+                               zend_array_destroy(op_array->dynamic_func_defs[i]->static_variables);
+                               op_array->dynamic_func_defs[i]->static_variables = NULL;
+                       }
                        destroy_op_array(op_array->dynamic_func_defs[i]);
                }
                efree(op_array->dynamic_func_defs);
        }
 }
 
 static void zend_update_extended_stmts(zend_op_array *op_array)
```

This patch fixes all the memory leaks for "Zend" tests. May be, it needs additional check for ZEND_ACC_CLOSURE.

Reply:
I added an extra test to show the need for the `ZEND_ACC_CLOSURE` check in https://github.com/php/php-src/pull/5595/commits/a9ff74f40e877612e4888a2101c6ede4d1a2ebf5.

With that check, I think doing this is safe, because we know that no new closure can be created from the prototype function anymore after that point, so the static variables will not be needed anymore. I'm not completely sure this is the better approach.

Reply:
Okay, I went with your suggestion.